sdk/typescript/src/experimental/types.ts

```diff
@@ -0,0 +1,253 @@
+        return writer.writeVec(bytes, (writer, el) => writer.write8(el));
+      },
+      (reader) => {
+        let bytes = reader.readVec((reader) => reader.read8());
+        return encodeStr(new Uint8Array(bytes), 'base64');
+      }
+    )
+);
+
+export type SuiObjectRef = {
+  objectId: string;
+  version: number;
+  digest: string;
+}
+
+initializers.push((b: typeof BCS) =>
+  b.registerStructType("SuiObjectRef", {
+    objectId: "ObjectID",
+    version: "u64",
+    digest: "ObjectDigest",
+  })
+);
+
+/**
+ * Transaction type used for transfering Coin objects between accounts.
+ * For this transaction to be executed, and `SuiObjectRef` should be queried
+ * upfront and used as a parameter.
+ */
+export type TransferCoinTx = {
+  TransferCoin: {
+    recipient: string;
+    object_ref: SuiObjectRef;
+  };
+}
+
+initializers.push((b: typeof BCS) =>
+  b.registerStructType("TransferCoinTx", {
+    recipient: "SuiAddress",
+    object_ref: "SuiObjectRef",
+  })
+);
+
+/**
+ * Transaction type used for publishing Move modules to the Sui.
+ * Should be already compiled using `sui-move`, example:
+ * ```
+ * $ sui-move build
+ * $ cat build/project_name/bytecode_modules/module.mv
+ * ```
+ * In JS:
+ * ```
+ * let file = fs.readFileSync('./move/build/project_name/bytecode_modules/module.mv');
+ * let bytes = Array.from(bytes);
+ * let modules = [ bytes ];
+ *
+ * // ... publish logic ...
+ * ```
+ *
+ * Each module should be represented as a sequence of bytes.
+ */
+export type PublishTx = {
+  Publish: {
+    modules: Iterable<Iterable<number>>;
+  };
+}
+
+initializers.push((b: typeof BCS) =>
+  b.registerStructType("PublishTx", {
+    modules: "vector<vector<u8>>",
+  })
+);
+
+// ========== Move Call Tx ===========
+
+/**
+ * An argument for the transaction. It is a 'meant' enum which expects to have
+ * one of the optional properties. If not, the BCS error will be thrown while
+ * attempting to form a transaction.
+ *
+ * Example:
+ * ```js
+ * let arg1: CallArg = { SharedObject: '5460cf92b5e3e7067aaace60d88324095fd22944' };
+ * let arg2: CallArg = { Pure: bcs.ser('u64', 100000) };
+ * let arg3: CallArg = { ImmOrOwnedObject: {
+ *   ObjectID: '4047d2e25211d87922b6650233bd0503a6734279',
+ *   SequenceNumber: 1,
+ *   ObjectDigest: Buffer.from('bCiANCht4O9MEUhuYjdRCqRPZjr2rJ8MfqNiwyhmRgA=', 'base64')
+ * } };
+ * ```
+ *
+ * For `Pure` arguments BCS is required. You must encode the values with BCS according
+ * to the type required by the called function. Pure accepts only serialized values
+ */
+export type CallArg =
+  | { SharedObject: string }
+  | { Pure: Iterable<number> }
+  | { ImmOrOwnedObject: SuiObjectRef }
+
+initializers.push((b: typeof BCS) =>
+  b.registerEnumType("CallArg", {
+    Pure: "vector<u8>",
+    ImmOrOwnedObject: "SuiObjectRef",
+    SharedObject: "ObjectID",
+  })
+);
+
+export type StructTag = {
+  address: string,
+  module: string,
+  name: string,
+  typeParams: TypeTag[]
+};
+
+export type TypeTag =
```

Comment:
should we have this in the non-experimental section ? this seems to mirror what we have in the rust side already.